### PR TITLE
Add additional validation for Query transfer transactions

### LINKED ISSUE #332: Queries: Additional validation on transfer transactions that pay for queries.
**Summary**
Current clients can submit queries which are paid for by transfer transactions.
Currently we validate the payer signature and balance on the transfer transactions.
However a client could submit a query with a transfer transaction that could pay the network fees but NOT cover the node's fee to satisfy the query.

**Possible resolution**
We can protect against this by adding the following validations to query payments 

- [x] Validate each payer in transferList has balance to cover the transfer amount.

- [x] If the transaction payerID also exists in transferList for query payment transfer transaction, validate that payer has balance to cover transaction fee and the transfer amount

- [x] Number of beneficiaries in transfer need not be one. 

- [x] But node should be getting required `queryFee` amount which is positive number

Unit tests to be done :

- [x]  multiple payers pay amount to node 

- [x] multiple payers pay amount to node as well as one more beneficiary. But node gets correct query payment fee

- [x] multiple payers pay amount to node as well as one more beneficiary. But node gets less query payment fee

- [x] TransactionPayer will pay for query payment to node and payer has enough balance

- [x] TransactionPayer will pay for query payment to node and payer has less balance

- [x] Transaction payer is not involved in transfers for query payment to node and all payers have enough balance

- [x] Transaction payer is not involved in transfers for query payment to node and one or more have less balance

### PR DESCRIPTION
**Pull request by Neeharika-Sompalli**
Closes #332 

Added additional validation for query payment transfer transactions
1. `Transfer list` of query payment transfer transaction should have only two account amounts in which payer account is transferring to node account.
2. `PayerAccountID` in transaction should be the same as the account that is sending amount in transfers list.
3. `nodeAccountID` should be the same as account that is amount receiving amount in transfers list.
4. Payer has balance to cover the transfer transaction and the transfer amount

Regression Passed:
https://hedera-hashgraph.slack.com/archives/C018CBQBTGB/p1600287191360500

### CODE REVIEW

File: test-clients/src/main/java/com/hedera/services/bdd/spec/queries/HapiQueryOp.java [hunk 1]
```diff
@@ -170,7 +172,9 @@ protected long feeFor(HapiApiSpec spec, Transaction txn, int numPayerKeys) throw
 	}
 
 	private Transaction fittedPayment(HapiApiSpec spec) throws Throwable {
-		if (nodePaymentFn.isPresent()) {
+		if (explicitPayment.isPresent()) {
+			return explicitPayment.get().signedTxnFor(spec);
```
Comment (ljianghedera) [inline test-clients/src/main/java/com/hedera/services/bdd/spec/queries/HapiQueryOp.java:176]: A silly question, queryHeader has a payment field of Transaction type. Is this explicitPayment use that field or it's a separate entity? How is the explicit payment transaction sent to services side if it's separate?

File: hedera-node/src/main/java/com/hedera/services/queries/validation/QueryFeeCheck.java [hunk 2]
```diff
@@ -117,4 +121,89 @@ ResponseCodeEnum adjustmentPlausibility(AccountAmount adjustment) {
 
 		return OK;
 	}
+
```
Comment (ljianghedera) [inline hedera-node/src/main/java/com/hedera/services/queries/validation/QueryFeeCheck.java:124]: From line 124-209, not sure if there is any special purpose for this or not. If this simply for validate the payment transaction, I remember in the AnswerFlow class, we should already have a interface to fulfill this functionality.

Comment (Neeharika-Sompalli) [reply hedera-node/src/main/java/com/hedera/services/queries/validation/QueryFeeCheck.java:124]: @ljianghedera If we don't specify `explicitPayment` a default transaction is generated. Else it uses the payment transaction we provide.

File: test-clients/src/main/java/com/hedera/services/bdd/spec/queries/HapiQueryOp.java (hunk 1 above)
Comment (Neeharika-Sompalli) [reply test-clients/src/main/java/com/hedera/services/bdd/spec/queries/HapiQueryOp.java:176]: @ljianghedera If we don't specify `explicitPayment` a default transaction is generated. Else it uses the payment transaction we provide.

File: hedera-node/src/main/java/com/hedera/services/queries/validation/QueryFeeCheck.java (hunk 2 above)
Comment (Neeharika-Sompalli) [reply hedera-node/src/main/java/com/hedera/services/queries/validation/QueryFeeCheck.java:124]: @ljianghedera `StakedAnswerFlow-> satisfyUsing->validatePayment`  calls  the same method `validateTransactionPreConsensus` , to validate transactions before consensus. 

These are extra validations needed only for query payment transactions. So, apart from the validations done in `validateTransactionFeeCoverage` additional validations are added in `queryFeeCheck.validateQueryPaymentTransaction`

**Review by ljianghedera — APPROVED** [review]
LGTM.

**Pull request closed without merging**